bblfsh/pyuast.cc

```diff
@@ -1,272 +1,760 @@
 #include <cstdint>
 #include <cstdlib>
 #include <cstring>
+#include <map>
 
 #include <Python.h>
+#include <structmember.h>
 
-#include "uast.h"
-#include "memtracker.h"
-
-
+#include "libuast.hpp"
 
 // Used to store references to the Pyobjects instanced in String() and
 // ItemAt() methods. Those can't be DECREF'ed to 0 because libuast uses them
 // so we pass ownership to these lists and free them at the end of filter()
-MemTracker memTracker;
 
-// WARNING: calls to Attribute MUST Py_DECREF the returned value once
-// used (or add it to the memtracker)
-static PyObject *Attribute(const void *node, const char *prop) {
-  PyObject *n = (PyObject *)node;
-  return PyObject_GetAttrString(n, prop);
+PyObject* asPyBuffer(uast::Buffer buf) {
+    PyObject* arr = PyByteArray_FromStringAndSize((const char*)(buf.ptr), buf.size);
+    free(buf.ptr);
+    return arr;
+    //return PyMemoryView_FromMemory((char*)(buf.ptr), buf.size, PyBUF_READ);
 }
 
-// WARNING: calls to AttributeValue MUST Py_DECREF the returned value once
-// used (or add it to the memtracker)
-static PyObject *AttributeValue(const void *node, const char *prop) {
-  PyObject *a = Attribute(node, prop);
-  return a && a != Py_None ? a : NULL;
+/*
+static bool checkError(const Uast* ctx) {
+    char *error = LastError((Uast*)ctx);
+    if (!error) {
+        return true;
+    }
+    PyErr_SetString(PyExc_RuntimeError, error);
+    free(error);
+    return false;
 }
+*/
 
-static bool HasAttribute(const void *node, const char *prop) {
-  PyObject *o = AttributeValue(node, prop);
-  if (o == NULL) {
-    return false;
-  }
+bool isContext(PyObject* obj);
 
-  Py_DECREF(o);
-  return true;
+bool assertNotContext(PyObject* obj) {
+    if (isContext(obj)) {
+        PyErr_SetString(PyExc_RuntimeError, "cannot use uast context as a node");
+        return false;
+    }
+    return true;
 }
 
-static const char *String(const void *node, const char *prop) {
-  const char *retval = NULL;
-  PyObject *o = Attribute(node, prop);
-  if (o != NULL) {
-    retval = PyUnicode_AsUTF8(o);
-    memTracker.TrackItem(o);
-  }
-  return retval;
-}
+// ==========================================
+//  External UAST Node (managed by libuast)
+// ==========================================
 
-static size_t Size(const void *node, const char *prop) {
-  size_t retval = 0;
-  PyObject *o = Attribute(node, prop);
-  if (o != NULL) {
-    retval = PySequence_Size(o);
-    Py_DECREF(o);
-  }
+class ContextExt;
 
-  return retval;
-}
+typedef struct {
+  PyObject_HEAD
+  ContextExt    *ctx;
```

Comment:
Probably a nit pick, but is there any particular reason for this formatting?

Reply:
Fixed in next commit.